number_theory/sieve.h

```diff
@@ -0,0 +1,97 @@
+#ifndef NUMBER_THEORY_SIEVE_H_
+#define NUMBER_THEORY_SIEVE_H_
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <bit>
+#include <limits>
+#include <stdexcept>
+#include <vector>
+
+#include "number_theory/utility.h"
+
+namespace tql {
+namespace number_theory {
+
+// Sieve of Euler.
+// It finds all prime numbers under a certain limit.
+// It also provides the factorizations of all numbers under the limit.
+//
+// The numbers are of type |T|, but they are converted to type |U| during
+// multiplication to avoid overflow.
+template <typename T, typename U = uint64_t>
+class Sieve {
+  static_assert(std::numeric_limits<T>::is_integer &&
+                    std::numeric_limits<U>::is_integer,
+                "Sieve must use integer types.");
+
+ public:
+  // The type of numbers used by the Sieve.
+  using type = T;
+
+  // Constructs the Sieve in linear time.
+  explicit Sieve(T num_limit)
+      : num_limit_(std::move(num_limit)), min_prime_factor_(num_limit_, T(0)) {
```

Comment:
I think it is a common knowledge that the default constructor for numbers initializes them to 0. So, I would avoid being explicit on them. For example:
```cpp
// bad/explicit due to being obvious
string s = "";
vector<bool> vb(n, false);
vector<double> vd(n, 0.0);
vector<int> vi(n, 0);

// good/implicit
string s;
vector<bool> vb(n);
vector<double> vd(n);
vector<int> vi(n);
```